File: app/src/main/java/com/kacper/krakgo/screens/sign_in/register_part2/RegisterPart2Presenter.java
package com.kacper.krakgo.screens.sign_in.register_part2;

import android.content.Context;
import android.net.Uri;
import android.support.annotation.NonNull;

import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.OnFailureListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.auth.UserProfileChangeRequest;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.storage.FirebaseStorage;
import com.google.firebase.storage.StorageReference;
import com.google.firebase.storage.UploadTask;
import com.kacper.krakgo.KrakGoApp;
import com.kacper.krakgo.helpers.FirebaseDatabaseHelper;
import com.kacper.krakgo.model.UserDetails;

/**
 * Created by kacper on 27/10/2017.
 */

public class RegisterPart2Presenter implements RegisterPart2Contract.Presenter {

    private RegisterPart2Contract.View mView;
    private Context mContext;

    public RegisterPart2Presenter(RegisterPart2Contract.View mView, Context mContext) {
        this.mView = mView;
        this.mContext = mContext;
    }

    @Override
    public void saveUserDetails(Uri mPhotoUri) {
        FirebaseUser user = KrakGoApp.getCurrentUser();
        FirebaseStorage storage = FirebaseStorage.getInstance();
        StorageReference storageRef = storage.getReference();
        StorageReference reference = storageRef.child("avatars/" + user.getUid());

        UploadTask uploadTask = reference.putFile(mPhotoUri);
        uploadTask
                .addOnCompleteListener(new OnCompleteListener<UploadTask.TaskSnapshot>() {
                    @Override
                    public void onComplete(@NonNull Task<UploadTask.TaskSnapshot> task) {
                        mView.photoUploadFinished(task.getResult().getDownloadUrl());
                    }
                })
                .addOnFailureListener(new OnFailureListener() {
                    @Override
                    public void onFailure(@NonNull Exception e) {
                        mView.showError(e);
                    }
                });
    }

    @Override
    public void updateUserProfile(UserDetails userDetails) {
        final FirebaseUser user = KrakGoApp.getCurrentUser();
        userDetails.setUserID(user.getUid());
        final UserProfileChangeRequest profileChangeRequest = new UserProfileChangeRequest.Builder()
                .setDisplayName(userDetails.getDisplay_name())
                .setPhotoUri(Uri.parse(userDetails.getPhoto_url()))
                .build();

        DatabaseReference database = KrakGoApp.getFirebaseReference();
        database.child(FirebaseDatabaseHelper.USER_DETAILS)
                .child(user.getUid())
                .setValue(userDetails)
                .addOnCompleteListener(new OnCompleteListener<Void>() {
                    @Override
                    public void onComplete(@NonNull Task<Void> task) {
                        user.updateProfile(profileChangeRequest).addOnCompleteListener(new OnCompleteListener<Void>() {
                            @Override
                            public void onComplete(@NonNull Task<Void> task) {
                                mView.userDetailsUpdated();
                            }
                        });
                    }
                })
                .addOnFailureListener(new OnFailureListener() {
                    @Override
                    public void onFailure(@NonNull Exception e) {
                        mView.showError(e);
                    }
                });
    }


}

File: app/src/main/java/com/kacper/krakgo/screens/sign_in/register_part2/RegisterPart2Presenter.kt
package com.kacper.krakgo.screens.sign_in.register_part2

import android.content.Context
import android.net.Uri

import com.google.android.gms.tasks.OnCompleteListener
import com.google.firebase.auth.UserProfileChangeRequest
import com.google.firebase.storage.FirebaseStorage
import com.kacper.krakgo.KrakGoApp
import com.kacper.krakgo.helpers.FirebaseDatabaseHelper
import com.kacper.krakgo.model.UserDetails
import com.kacper.krakgo.mvp.MvpPresenterImpl
import com.kacper.krakgo.screens.home.map.MapContract

/**
 * Created by kacper on 27/10/2017.
 */

class RegisterPart2Presenter : MvpPresenterImpl<RegisterPart2Contract.View>(),
            RegisterPart2Contract.Presenter {

    override fun saveUserDetails(mPhotoUri: Uri) {

        val storage = FirebaseStorage.getInstance()
        val storageRef = storage.reference
        val reference = storageRef.child("avatars/" + getCurrentUser()?.uid)

        val uploadTask = reference.putFile(mPhotoUri)
        uploadTask
                .addOnCompleteListener { task -> mView?.photoUploadFinished(task.result.downloadUrl!!) }
                .addOnFailureListener { e -> mView?.showError(e) }
    }

    override fun updateUserProfile(userDetails: UserDetails) {
        userDetails.userID = getCurrentUser()?.uid
        val profileChangeRequest = UserProfileChangeRequest.Builder()
                .setDisplayName(userDetails.display_name)
                .setPhotoUri(Uri.parse(userDetails.photo_url))
                .build()

        getDatabaseReference().child(FirebaseDatabaseHelper.USER_DETAILS)
                .child(getCurrentUser()?.uid)
                .setValue(userDetails)
                .addOnCompleteListener { getCurrentUser()?.updateProfile(profileChangeRequest)
                        ?.addOnCompleteListener({ mView?.userDetailsUpdated() }) }
                .addOnFailureListener { e -> mView?.showError(e) }
    }


}
